### PR TITLE
Productionize CurrentShardRegionState #27406

### PR DESCRIPTION
**Pull request by helena**
https://github.com/akka/akka/issues/27406, similar to https://github.com/akka/akka/pull/27395

### CODE REVIEW

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 1]
```diff
@@ -367,7 +367,8 @@ object ShardRegion {
    *
    * If gathering the shard information times out the set of shards will be empty.
    */
-  @SerialVersionUID(1L) final case class CurrentShardRegionState(shards: Set[ShardState]) {
+  @SerialVersionUID(1L) final case class CurrentShardRegionState(val shards: Set[ShardState], val failed: Set[ShardId])
```
Comment (helena) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: Because this has no bin compat issue, having not been serialized before, I elected to not convert to a class to keep it simple.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGetStateSpec.scala [hunk 2]
```diff
@@ -61,12 +61,6 @@ object ClusterShardingGetStateSpecConfig extends MultiNodeConfig {
       dir = target/ClusterShardingGetStateSpec/sharding-ddata
       map-size = 10 MiB
     }
-    # using Java serialization for these messages because test is sending them
-    # to other nodes, which isn't normal usage.
-    akka.actor.serialization-bindings {
-      "${ShardRegion.GetShardRegionState.getClass.getName}" = java-test
-      "${classOf[ShardRegion.CurrentShardRegionState].getName}" = java-test
-    }
```
Comment (helena) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGetStateSpec.scala]: Now this is in the serializer, no longer needed.

**Review by raboof — COMMENTED** [review]
Nice improvement. The tests are a bit hard to follow but I guess that was already the case.

Weird that mima didn't catch that unapply incompatibility...

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingQueriesSpec.scala [hunk 3]
```diff
@@ -146,13 +146,32 @@ abstract class ClusterShardingQueriesSpec
         val timeouts = numberOfShards / regions.size
 
         // 3 regions, 2 shards per region, all 2 shards/region were unresponsive
-        // within shard-region-query-timeout = 0ms
-        regions.values.forall { s =>
-          s.stats.isEmpty && s.failed.size == timeouts
-        } shouldBe true
+        // within shard-region-query-timeout = 0ms on all but first
```
[the remark this answers is not in the document]
Comment (helena) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingQueriesSpec.scala]: Yes only the busy node leverages 0ms timeout in order to test the some failed while others responded - where before the related change it was all empty - regardless if some had responded.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 4]
```diff
@@ -367,7 +362,13 @@ object ShardRegion {
    *
    * If gathering the shard information times out the set of shards will be empty.
    */
-  @SerialVersionUID(1L) final case class CurrentShardRegionState(shards: Set[ShardState]) {
+  @SerialVersionUID(1L) final case class CurrentShardRegionState(shards: Set[ShardState], failed: Set[ShardId])
```
[the remark this answers is not in the document]
Comment (helena) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: Sure, I did start this change out by doing that, then reverted to the original case class thinking because it was not yet in the serializer and sent over the wire it may not need it. I"ll make the change, thanks.

[the remark this answers is not in the document]
Comment (raboof) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: I figured out why mima missed this: the fix at https://github.com/lightbend/mima/pull/468/files#diff-1b17b4fbc8ceecfa206a29f9a469ff3aR21 is not yet released.

[the remark this answers is not in the document]
Comment (raboof) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: verified with a snapshot from mima master, there it does catch this problem and successfully verifies the latest version of this branch is OK :+1:

**Review by raboof — APPROVED** [review]

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingQueriesSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingQueriesSpec.scala]: So it's the opposite of this comment, 0ms on first/busy but not on second and third?

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by helena** [comment]
@johanandren ya, I updated the test comment to now correctly say
https://github.com/akka/akka/pull/28645/commits/9eaab84b93a40ec6734efed4d52632e729825f6e#diff-ddde478ff8e81b71fc455ddf3bba6a7bR149

**Comment by helena** [comment]
With this now we can do this https://github.com/akka/akka-management/issues/577

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (helena) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: Thanks!
